### ISSUE TITLE
proposal: `do_not_use_setters_of_arguments`

### ISSUE THREAD

**Issue by minhqdao** [dart-lang/linter#4081]
## Description

Do not use setters of passed function arguments.

## Details

I recently saw someone doing the following:

```dart
Future<void> loadEvents(User user) async {
  final List<Event> events = await source.loadEvents();
  user.events = events;
}
```

Using data which is stored in arguments is, of course, normal:

```dart
  final List<Event> events = await source.loadEvents(uid: user.id);
```

But I have a strong feeling of discomfort looking at the first code block. Or is it just me?

The code did work because `user` is a reference to the variable declared in the scope of the calling code. But I would personally much prefer the following because here we are clearly separating the "getting" from the "setting":

```dart
Future<List<Event>> loadEvents() async {
  final <List<Event>> events = await source.loadEvents();
  return events;
}

final List<Event> events = loadEvents();
user.events = events;
```

My question is: Could this be a general rule that we could/should create a lint for or am I missing a case where this pattern is valid or even wanted? Or is it just a matter of taste?

## Kind

Coding style.

### Discussion checklist

This is strongly related to and complements `parameter_assignments`.

- [x] List any existing rules this proposal modifies, complements, overlaps or conflicts with.
- [x] List any relevant issues (reported here, the [SDK Tracker], or elsewhere).
- [x] If there's any prior art (e.g., in other linters), please add references here.
- [x] If this proposal corresponds to [Effective Dart] or [Flutter Style Guide] advice, please call it out. (If there isn't any corresponding advice, should there be?)
- [x] If this proposal is motivated by real-world examples, please provide as many details as you can.  Demonstrating potential impact is _especially valuable_.

[Writing Lints]: https://github.com/dart-lang/linter/blob/main/doc/writing-lints.md
[Effective Dart]: https://dart.dev/guides/language/effective-dart
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[SDK Tracker]: https://github.com/dart-lang/sdk/issues
